File: leetcode/typescript/0001-two-sum.ts
/** https://leetcode.com/problems/two-sum
 * @param {number[]} nums
 * @param {number} target
 * @returns {number[]}
 */
function twoSum(nums: number[], target: number): number[] {
  const numMap = new Map<number, number>();

  for (let i = 0; i < nums.length; i++) {
    const numIdx = numMap.get(target - nums[i]);

    if (numIdx !== undefined) {
      return [numIdx, i];
    }

    numMap.set(nums[i], i);
  }
  return [];
}

console.log(twoSum([2, 7, 11, 15], 9));
console.log(twoSum([3, 2, 4], 6));
console.log(twoSum([3, 3], 6));

File: leetcode/cpp/0001-two-sum.cpp
#include <unordered_map>
#include <vector>

using std::unordered_map;
using std::vector;

class Solution {
public:
    vector<int> twoSum(vector<int> &nums, int target) {
        unordered_map<int, int> numMap;
        for (int i = 0; i < static_cast<int>(nums.size()); i++) {
            int num{nums[static_cast<size_t>(i)]};
            int complement{target - num};
            if (numMap.count(complement)) {
                return {numMap[complement], i};
            }
            numMap[num] = i;
        }

        return {};
    }
};
